src/pylibsshext/sftp.pyx

```diff
@@ -54,69 +57,186 @@ cdef class SFTP:
             self._libssh_sftp_session = NULL
 
     def put(self, local_file, remote_file):
+        SFTP_AIO(self).put(local_file, remote_file)
+
+    def get(self, remote_file, local_file):
+        SFTP_AIO(self).get(remote_file, local_file)
+
+    def close(self):
+        if self._libssh_sftp_session is not NULL:
+            sftp.sftp_free(self._libssh_sftp_session)
+            self._libssh_sftp_session = NULL
+
+    def _get_sftp_error_str(self):
+        error = sftp.sftp_get_error(self._libssh_sftp_session)
+        if error in MSG_MAP and error != sftp.SSH_FX_FAILURE:
+            return MSG_MAP[error]
+        return "Generic failure: %s" % self.session._get_session_error_str()
+
+cdef sftp.sftp_session get_sftp_session(SFTP sftp_obj):
+    return sftp_obj._libssh_sftp_session
+
+cdef class SFTP_AIO:
+    def __cinit__(self, SFTP sftp_obj):
+        self._sftp = get_sftp_session(sftp_obj)
+
+        self._limits = sftp.sftp_limits(self._sftp)
+        if self._limits is NULL:
+            raise LibsshSFTPException("Failed to get remote SFTP limits [%s]" % (self._get_sftp_error_str()))
+
+    def __init__(self, SFTP sftp_obj):
+        self._aio_queue = []
+
+    def __dealloc__(self):
+        if self._rf is not NULL:
+            sftp.sftp_close(self._rf)
+            self._rf = NULL
+
+    def put(self, local_file, remote_file):
+        # reset
+        self._aio_queue = []
+        self._total_bytes_requested = 0
+
+        cdef C_AIO aio
         cdef sftp.sftp_file rf
-        with open(local_file, "rb") as f:
-            remote_file_b = remote_file
-            if isinstance(remote_file_b, unicode):
-                remote_file_b = remote_file.encode("utf-8")
+        self._remote_file = remote_file
 
-            rf = sftp.sftp_open(self._libssh_sftp_session, remote_file_b, O_WRONLY | O_CREAT | O_TRUNC, sftp.S_IRWXU)
-            if rf is NULL:
-                raise LibsshSFTPException("Opening remote file [%s] for write failed with error [%s]" % (remote_file, self._get_sftp_error_str()))
-            buffer = f.read(1024)
-
-            while buffer != b"":
-                length = len(buffer)
-                written = sftp.sftp_write(rf, PyBytes_AS_STRING(buffer), length)
-                if written != length:
-                    sftp.sftp_close(rf)
+        remote_file_b = remote_file
+        if isinstance(remote_file_b, unicode):
+            remote_file_b = remote_file.encode("utf-8")
+
+        rf = sftp.sftp_open(self._sftp, remote_file_b, O_WRONLY | O_CREAT | O_TRUNC, sftp.S_IRWXU)
+        if rf is NULL:
+            raise LibsshSFTPException("Opening remote file [%s] for write failed with error [%s]" % (remote_file, self._get_sftp_error_str()))
+        self._rf = rf
+
+        with open(local_file, "rb") as f:
+            f.seek(0, os.SEEK_END)
+            self._file_size = f.tell()
+            f.seek(0, os.SEEK_SET)
+
+            # open up to 10 parallel transfers
```

Comment:
It would be interesting to look into threading in the context of Python 3.13t. But that's obviously out of the scope here. I do agree that this word should be replaced, though.

Reply:
Right. I do not expect treading would help the sftp performance thoughput as it needs to go through a single tcp connection and ssh session (and sharing ssh session among different threads is not supported by libssh).